File: xls/dslx/scanner_test.py
# Lint as: python3
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for xls.dslx.scanner."""

from typing import Text, Sequence, Any

from xls.dslx import scanner
from xls.dslx.scanner import TokenKind
from absl.testing import absltest


class ScannerTest(absltest.TestCase):

  def assertLen(self, sequence: Sequence[Any], target: int) -> None:
    assert len(sequence) == target, 'Got sequence of length %d, want %d' % (
        len(sequence), target)

  def make_scanner(self, text: Text, **kwargs) -> scanner.Scanner:
    return scanner.Scanner('<fake>', text, **kwargs)

  def test_scan_keyword(self):
    s = self.make_scanner('fn')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_keyword(scanner.Keyword.FN))

  def test_function_definition(self):
    s = self.make_scanner('fn ident(x) { x }')
    tokens = s.pop_all()

    self.assertTrue(
        tokens[0].is_keyword(scanner.Keyword.FN), msg=repr(tokens[0]))
    self.assertEqual('fn', str(tokens[0]))
    self.assertTrue(tokens[1].is_identifier('ident'))
    self.assertEqual('ident', str(tokens[1]))
    self.assertEqual(tokens[2].kind, TokenKind('('))
    self.assertEqual('OPAREN', str(tokens[2]))
    self.assertTrue(tokens[3].is_identifier('x'))
    self.assertEqual('x', str(tokens[3]))
    self.assertEqual(tokens[4].kind, TokenKind(')'))
    self.assertEqual('CPAREN', str(tokens[4]))
    self.assertEqual(tokens[5].kind, TokenKind('{'))
    self.assertEqual('OBRACE', str(tokens[5]))
    self.assertTrue(tokens[6].is_identifier('x'))
    self.assertEqual('x', str(tokens[6]))
    self.assertEqual(tokens[7].kind, TokenKind('}'))
    self.assertEqual('CBRACE', str(tokens[7]))

  def test_doubled_up_plus(self):
    s = self.make_scanner('fn concat(x, y) { x ++ y }')
    tokens = s.pop_all()

    self.assertTrue(tokens[-4].is_identifier('x'))
    self.assertEqual(tokens[-3].kind, TokenKind('++'))
    self.assertTrue(tokens[-2].is_identifier('y'))
    self.assertEqual(tokens[-1].kind, TokenKind('}'))

  def test_number_hex(self):
    s = self.make_scanner('0xf00')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('0xf00'), msg=str(tokens))

  def test_negative_number_hex(self):
    s = self.make_scanner('-0xf00')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('-0xf00'), msg=str(tokens))

  def test_number_bin(self):
    s = self.make_scanner('0b10')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('0b10'), msg=str(tokens))

    with self.assertRaisesRegex(scanner.ScanError,
                                'Invalid digit for binary number: 2'):
      s = self.make_scanner('0b102')
      tokens = s.pop_all()

  def test_negative_number_bin(self):
    s = self.make_scanner('-0b10')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('-0b10'), msg=str(tokens))

  def test_negative_number(self):
    s = self.make_scanner('-42')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('-42'), msg=str(tokens))

  def test_number_with_underscores(self):
    s = self.make_scanner('0b11_1100')
    tokens = s.pop_all()
    self.assertLen(tokens, 1)
    self.assertTrue(tokens[0].is_number('0b11_1100'), msg=str(tokens))

  def test_pos_lt(self):
    self.assertLess(scanner.Pos('<fake>', 0, 0), scanner.Pos('<fake>', 0, 1))
    self.assertLess(scanner.Pos('<fake>', 0, 0), scanner.Pos('<fake>', 1, 0))
    self.assertGreaterEqual(
        scanner.Pos('<fake>', 0, 0), scanner.Pos('<fake>', 0, 0))

  def test_scan_incomplete_number(self):
    with self.assertRaisesRegex(scanner.ScanError, 'Expected hex characters'):
      self.make_scanner('0x').pop_all()
    with self.assertRaisesRegex(scanner.ScanError,
                                'Expected binary characters'):
      self.make_scanner('0b').pop_all()

  def test_scan_incomplete_character(self):
    with self.assertRaisesRegex(scanner.ScanError,
                                'Expected closing single quote'):
      self.make_scanner("'a").pop_all()
    with self.assertRaisesRegex(scanner.ScanError,
                                'Expected character after single quote'):
      self.make_scanner("'").pop_all()

  def test_scan_in_whitespace_and_comments_mode(self):
    program = """// Hello comment world.
    42
  // EOF"""
    tokens = self.make_scanner(
        program, include_whitespace_and_comments=True).pop_all()
    self.assertLen(tokens, 5)
    self.assertEqual(tokens[0].kind, TokenKind.COMMENT)
    self.assertEqual(tokens[1].kind, TokenKind.WHITESPACE)
    self.assertEqual(tokens[2].kind, TokenKind.NUMBER)
    self.assertEqual(tokens[3].kind, TokenKind.WHITESPACE)
    self.assertEqual(tokens[4].kind, TokenKind.COMMENT)

  def test_peek_pop_drop_try_drop(self):
    text = '[!](-)'
    expected = [
        TokenKind.OBRACK,
        TokenKind.BANG,
        TokenKind.CBRACK,
        TokenKind.OPAREN,
        TokenKind.MINUS,
        TokenKind.CPAREN,
    ]
    s = self.make_scanner(text)
    for i, tk in enumerate(expected):
      self.assertFalse(s.at_eof())
      self.assertEqual(s.peek().kind, tk)
      if i % 2 == 0:
        t = s.pop()
        self.assertIsInstance(t, scanner.Token)
        self.assertEqual(t.kind, tk)
      else:
        self.assertTrue(s.try_pop(tk))
    self.assertTrue(s.at_eof())

if __name__ == '__main__':
  absltest.main()

File: xls/dslx/scanner_test.cc
// Copyright 2020 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "xls/dslx/scanner.h"

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/random/random.h"
#include "xls/common/status/matchers.h"

namespace xls::dslx {
namespace {

using status_testing::StatusIs;
using testing::HasSubstr;

static absl::StatusOr<std::vector<Token>> ToTokens(std::string text) {
  Scanner s("fake_file.x", std::move(text));
  return s.PopAll();
}

TEST(ScannerTest, SimpleTokens) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("+ - ++ << >>"));
  ASSERT_EQ(5, tokens.size());
  EXPECT_EQ(tokens[0].kind(), TokenKind::kPlus);
  EXPECT_EQ(tokens[1].kind(), TokenKind::kMinus);
  EXPECT_EQ(tokens[2].kind(), TokenKind::kDoublePlus);
  EXPECT_EQ(tokens[3].kind(), TokenKind::kDoubleOAngle);
  EXPECT_EQ(tokens[4].kind(), TokenKind::kDoubleCAngle);
}

TEST(ScannerTest, HexNumbers) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens,
                           ToTokens("0xf00 0xba5 0xA"));
  ASSERT_EQ(3, tokens.size());
  EXPECT_TRUE(tokens[0].IsNumber("0xf00"));
  EXPECT_TRUE(tokens[1].IsNumber("0xba5"));
  EXPECT_TRUE(tokens[2].IsNumber("0xA"));
}

TEST(ScannerTest, BoolKeywords) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens,
                           ToTokens("true false bool"));
  ASSERT_EQ(3, tokens.size());
  EXPECT_TRUE(tokens[0].IsKeyword(Keyword::kTrue));
  EXPECT_TRUE(tokens[1].IsKeyword(Keyword::kFalse));
  EXPECT_TRUE(tokens[2].IsKeyword(Keyword::kBool));
  EXPECT_EQ(tokens[0].ToErrorString(), "keyword:true");
}

TEST(ScannerTest, IdentifierWithTick) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens,
                           ToTokens("state state' state'' s'"));
  ASSERT_EQ(4, tokens.size());
  EXPECT_TRUE(tokens[0].IsIdentifier("state"));
  EXPECT_TRUE(tokens[1].IsIdentifier("state'"));
  EXPECT_TRUE(tokens[2].IsIdentifier("state''"));
  EXPECT_TRUE(tokens[3].IsIdentifier("s'"));
}

TEST(ScannerTest, TickCannotStartAnIdentifier) {
  const char* kText = "'state";
  EXPECT_THAT(
      ToTokens(kText),
      StatusIs(
          absl::StatusCode::kInvalidArgument,
          HasSubstr(
              "Expected closing single quote for character literal; got t")));
}

// Verifies that Scanner::ProcessNextStringChar() correctly handles all
// supported escape sequences.
TEST(ScannerTest, RecognizesEscapes) {
  std::string text = R"(\n\r\t\\\0\'\"\x6f\u{fEdCb}Hello"extrastuff)";
  Scanner s("fake_file.x", text);
  XLS_ASSERT_OK_AND_ASSIGN(std::string result, s.ScanUntilDoubleQuote());
  EXPECT_EQ(static_cast<uint8_t>(result[0]), 10);    // Newline.
  EXPECT_EQ(static_cast<uint8_t>(result[1]), 13);    // Carriage return.
  EXPECT_EQ(static_cast<uint8_t>(result[2]), 9);     // Tab.
  EXPECT_EQ(static_cast<uint8_t>(result[3]), 92);    // Backslash.
  EXPECT_EQ(static_cast<uint8_t>(result[4]), 0);     // Null.
  EXPECT_EQ(static_cast<uint8_t>(result[5]), 39);    // Single quote.
  EXPECT_EQ(static_cast<uint8_t>(result[6]), 34);    // Double quote.
  EXPECT_EQ(static_cast<uint8_t>(result[7]), 111);   // Lowercase o.
  EXPECT_EQ(static_cast<uint8_t>(result[8]), 0xcb);  // Byte 3 of Unicode code.
  EXPECT_EQ(static_cast<uint8_t>(result[9]), 0xed);  // Byte 2 of Unicode code.
  EXPECT_EQ(static_cast<uint8_t>(result[10]), 0xf);  // Byte 1 of Unicode code.
  EXPECT_EQ(static_cast<uint8_t>(result[11]), 'H');  // Final word.
  EXPECT_EQ(static_cast<uint8_t>(result[12]), 'e');
  EXPECT_EQ(static_cast<uint8_t>(result[13]), 'l');
  EXPECT_EQ(static_cast<uint8_t>(result[14]), 'l');
  EXPECT_EQ(static_cast<uint8_t>(result[15]), 'o');
  EXPECT_EQ(result.size(), 16);
}

TEST(ScannerTest, ScanJustWhitespace) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens(" "));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_EQ(tokens[0].kind(), TokenKind::kEof);
}

TEST(ScannerTest, ScanKeyword) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("fn"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsKeyword(Keyword::kFn));
}

TEST(ScannerTest, FunctionDefinition) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens,
                           ToTokens("fn ident(x) { x }"));

  ASSERT_EQ(tokens.size(), 8);

  EXPECT_TRUE(tokens[0].IsKeyword(Keyword::kFn));
  EXPECT_EQ(tokens[0].ToString(), "fn");

  EXPECT_TRUE(tokens[1].IsIdentifier("ident"));
  EXPECT_EQ(tokens[1].ToString(), "ident");

  EXPECT_EQ(tokens[2].kind(), TokenKindFromString("(").value());
  EXPECT_EQ(tokens[2].ToString(), "(");

  EXPECT_TRUE(tokens[3].IsIdentifier("x"));
  EXPECT_EQ(tokens[3].ToString(), "x");

  EXPECT_EQ(tokens[4].kind(), TokenKindFromString(")").value());
  EXPECT_EQ(tokens[4].ToString(), ")");

  EXPECT_EQ(tokens[5].kind(), TokenKindFromString("{").value());
  EXPECT_EQ(tokens[5].ToString(), "{");

  EXPECT_TRUE(tokens[6].IsIdentifier("x"));
  EXPECT_EQ(tokens[6].ToString(), "x");

  EXPECT_EQ(tokens[7].kind(), TokenKindFromString("}").value());
  EXPECT_EQ(tokens[7].ToString(), "}");
}

TEST(ScannerTest, DoublePlus) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("x++y"));
  ASSERT_EQ(tokens.size(), 3);
  EXPECT_TRUE(tokens[0].IsIdentifier("x"));
  EXPECT_EQ(tokens[1].ToString(), "++");
  EXPECT_TRUE(tokens[2].IsIdentifier("y"));
}

TEST(ScannerTest, NumberHex) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("0xf00"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("0xf00"));
}

TEST(ScannerTest, NegativeNumberHex) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("-0xf00"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("-0xf00"));
}

TEST(ScannerTest, NumberBin) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("0b10"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("0b10"));
}

TEST(ScannerTest, NumberBinInvalidDigit) {
  EXPECT_THAT(ToTokens("0b102"),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Invalid digit for binary number: '2'")));
}

TEST(ScannerTest, NegativeNumberBin) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("-0b10"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("-0b10"));
}

TEST(ScannerTest, NegativeNumber) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("-42"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("-42"));
}

TEST(ScannerTest, NumberWithUnderscores) {
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, ToTokens("0b11_1100"));
  ASSERT_EQ(tokens.size(), 1);
  EXPECT_TRUE(tokens[0].IsNumber("0b11_1100"));
}

TEST(ScannerTest, ScanIncompleteNumbers) {
  EXPECT_THAT(ToTokens("0x"), StatusIs(absl::StatusCode::kInvalidArgument,
                                       HasSubstr("Expected hex characters")));
  EXPECT_THAT(ToTokens("0b"),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Expected binary characters")));
}

TEST(ScannerTest, BadlyFormedNumber) {
  EXPECT_THAT(ToTokens("u1:01"),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Invalid radix for number")));
}

TEST(ScannerTest, IncompleteCharacter) {
  EXPECT_THAT(ToTokens("'a"),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Expected closing single quote")));
  EXPECT_THAT(ToTokens("'"),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Expected character after single quote")));
}

TEST(ScannerTest, WhitespaceAndCommentsMode) {
  Scanner s("fake_file.x", R"(// Hello comment world.
  42
  // EOF)",
            /*include_whitespace_and_comments=*/true);
  XLS_ASSERT_OK_AND_ASSIGN(std::vector<Token> tokens, s.PopAll());
  EXPECT_EQ(tokens.size(), 5);
  EXPECT_EQ(tokens[0].kind(), TokenKind::kComment);
  EXPECT_EQ(tokens[1].kind(), TokenKind::kWhitespace);
  EXPECT_EQ(tokens[2].kind(), TokenKind::kNumber);
  EXPECT_EQ(tokens[3].kind(), TokenKind::kWhitespace);
  EXPECT_EQ(tokens[4].kind(), TokenKind::kComment);
}

TEST(ScannerTest, PopSeveral) {
  Scanner s("fake_file.x", "[!](-)");
  std::vector<TokenKind> expected = {
      TokenKind::kOBrack, TokenKind::kBang,  TokenKind::kCBrack,
      TokenKind::kOParen, TokenKind::kMinus, TokenKind::kCParen,
  };
  for (TokenKind tk : expected) {
    ASSERT_FALSE(s.AtEof());
    XLS_ASSERT_OK_AND_ASSIGN(Token t, s.Pop());
    EXPECT_EQ(t.kind(), tk);
  }
  EXPECT_TRUE(s.AtEof());
}

TEST(ScannerTest, ScanRandomLookingForCrashes) {
  absl::BitGen bitgen;
  for (int64_t i = 0; i < 256 * 1024; ++i) {
    int64_t length = absl::Uniform(bitgen, 0, 512);
    std::string text;
    for (int64_t charno = 0; charno < length; ++charno) {
      text.push_back(absl::Uniform(bitgen, 0, 256));
    }
    Scanner s("fake_file.x", text);
    absl::StatusOr<std::vector<Token>> tokens = s.PopAll();
    if (!tokens.ok()) {
      continue;
    }
    // Ensure any scanned tokens can be converted to strings.
    for (const Token& token : tokens.value()) {
      (void)token.ToString();
    }
  }
}

}  // namespace
}  // namespace xls::dslx
